### DISCUSSION TITLE
Max framerate

### DISCUSSION THREAD

**Discussion by Lukaswnd** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#411] [Ideas]
How hard would it be to implement a max framerate feature?
When you lower the PIXEL_COLOR_DEPTH_BITS and don't have a huge panel combination, you might endup with a very high framerate.
I know due to DMA there is not much overhead for a higher framerate. But on one hand it feels wrong have 400hz and on the other hand it should be possible to incease the OE time (time how long each row of leds is turned on) resulting in a brighter display.

Ofcourse you could reduce I2S speed, but I like to have my display brighter.

I currently have no idea how to implement this and I don't want to spend my time on this now, maybe someone else gives it a try.

**Comment by mrcodetastic** [comment]
The framerate is a function of the data bitrate (dma i2s speed). Using the lowest clock rate + highest panel brightness level should do the trick.

I am not keen on implementing any dynamic configuration on the DMA hardware side of the library based on color depth, resolution etc. as it's a pain to debug and adds unnecessary complexity.

**Reply by Lukaswnd** [reply]
Yeah, that is what I thought too much work for little result

**Comment by board707** [comment]
> it should be possible to increase the OE time (time how long each row of leds is turned on) resulting in a brighter display.

@mrfaptastic, correct me If I am wrong, but as I understand the framerate has nothing to do with panel brightness.
OE sets the fraction of time from the total cycle length in which the diodes are lit. OE time is a relative value, it does not depend on the duration of the cycle itself and therefore the framerate does not affect the brightness

**Reply by mrcodetastic** [reply]
That's correct. Upping the brightness will make the panel brighter though, so combined with the lower clock rate is the most pragmatic solution to extreme refresh rates.

**Reply by Lukaswnd** [reply]
I think you don't get exactly what I mean. Let me try to explain it a litte more precise.

Im just talking of a update for one row from, now on. And 1 frame updates 1 row (2 beauce of parrallel interface) and each frame is divided into subframes. I subframe has the duration of shifting all data into this row.
Let's assume a PIXEL_COLOR_DEPTH_BITS of 2. Then we have 5 subrames each frame. 
Subframe1: shift data of first bit (leds on)
Subframe2: show (oe) subframe 1
Subframe3: shift data of second bit (leds on)
Subframe4: show (oe) subframe 3
Subframe5: show (oe) subframe 3

This is how I understand this lib and these panels work.
The brightness is controlled by the time oe is actually active during the subframes. So if it is on all the time brightness is 100% and if it is on for just a few clks brightness is low. (There might be a brightness formula to map the mechanical brighness to the visual one.)

If you now scale up the number of oe subframes, which inceases the ratio of subframes whith leds on to the ones where they are off, the Panel appears brighter. 

But since one subframe has a fixed duration and the number of subframes in each frame increase the framerate lowers.

So what I thought was, if you limit the framerate, instead of always using the highes possibe, you are able to add some subframes inbetween each frame to incease the brightness.

By now I get that what I am actually looking for might be two totally different things. The one is to limit the framerate, so there is less cpu overhead. If it is solved by just delaying between each panel update would result in less brightness, because during the delay no leds would be on. The other thing  is to raise the number of subrames with glowing leds to maximise the brightness.

I guess the second one would be way easer to implement, even though i'm not sure if is worth it. The ratio is roughly:  (1/2) the framerate -> (3/2) the brightness

**Reply by mrcodetastic** [reply]
There's no CPU overhead of a faster frame rate. DMA engine runs completely separately from the CPU and pumps data directly out from the SRAM.

Either way, not sure this is worth the effort and extra complexity. Have already had a tonne of commits recently to fix the brightness code by another contributor.

**Reply by board707** [reply]
@Lukaswnd 
> Then we have 5 subrames each frame.
 Subframe1: shift data of first bit (leds on)
 Subframe2: show (oe) subframe 1
 Subframe3: shift data of second bit (leds on) 
 Subframe4: show (oe) subframe 3 
 Subframe5: show (oe) subframe 3

You do not quite correctly represent the work of the matrix. You assumed that at the moment of loading a next bit of data( Subframe1 and 3 in your scheme), the diodes on the matrix are turned off. But it is not.
In fact, the loading of a new subframe occurs DURING the exposure of the previous one:

 Subframe1: show (oe) bit1,  shift data of bit2 (leds on)
 Subframe2: show (oe) bit2 , shift data of bit3 (leds on)
 Subframe3: show (oe) bit2
 Subframe4: show (oe) bit3 , shift data of bit1 (leds on)
 Subframe5: show (oe) bit3
 Subframe6: show (oe) bit3
 Subframe7: show (oe) bit3

**Reply by mrcodetastic** [reply]
That makes sense.

Check out the 'brtCtrlOEv2' function as that was recently updated by another contributor to enhance the brightness control. Not really an area that I have touched upon to be honest.

**Reply by Lukaswnd** [reply]
@board707 oh I didn't know that. Thank you for explaining

**Reply by daveythacher** [reply]
The brightness loss is minimal when properly configured. At high refresh rates brightness reduction is expected. 

This code base does not support refresh rates about 240-250Hz. The frame rate can be up to this value however the CPU may limit this. Note I would limit to 120FPS.

**Reply by daveythacher** [reply]
Correction: This code base does not support refresh rates above 240-250Hz.
